### ISSUE TITLE
React 19 Support

### ISSUE THREAD

**Issue by florence-wolfe** [plouc/nivo#2618]
## Update (by @plouc)

- The allowed version range has been increased
- Some deprecated APIs have been removed (default props, prop-types, ...)
- Tried to use some nivo charts in a nextjs project with React 19, and it _worked_, had to use `overrides` for fixing the react version though, due to `react-spring`.

-----

**Is your feature request related to a problem? Please describe.**
- `@nivo/core` (and possibly other packages) have a strict React dependency range that restricts React 19
  - [@nivo/core's package.json](https://github.com/plouc/nivo/blob/master/packages/core/package.json#L38)
- This prevents applications from upgrading their React versions unless they begin including the `--legacy-peer-deps` flag or `--force` for every installation which is not ideal.

**Describe the solution you'd like**
- Either increase the allowing version range in the packages or add proper support for the version.
  - The React team has written a [migration guide](https://react.dev/blog/2024/04/25/react-19-upgrade-guide) that also includes several codemods to make the migration relatively seamless.

**Comment by plouc** [comment]
Adding support for React 19 might be a lot of work, it's also still unclear, looking at the migration guide, if I could still support older React versions with the same @nivo version.

The 19 still being a RC at the moment, I think it's not really pressing, and I'd probably need to upgrade gatsby, storybook... as well, which seems like a huge project.

Finally, I was migrating tests from `enzyme` (which has been deprecated) to use the react test renderer, which they're removing apparently, they advise to use `@testing-library/react`, but the approach of this lib is really not practical to test charts/svg elements.

It will take time to get there and this should be split in sub-projects, I don't have the bandwidth to handle everything at once.

**Comment by J4v4Scr1pt** [comment]
Will follow this thread, we are a company of the bigger size and are about to replace our current chart solution.
This was our top pick until I saw this thread.

I understand that time estimation is super hard, but do you think this will take years or months after React 19 release.
Because we (among others) use Next.Js and no React 19 support will also block us from upgrading nextjs version in the future.

**Comment by o-alexandrov** [comment]
React has [an issue](https://github.com/eps1lon/react/issues/10) where they link some libraries and respective PRs how those libraries upgraded to react 19.
It might be a useful read to familiarize with common roadblocks and how other devs resolved those in their projects.

**Comment by plouc** [comment]
@o-alexandrov, thank you for sharing, this could help.

**Comment by skip405** [comment]
Latest major version of Next.js (version 15) uses React 19 RC, so it's not going to be possible to use Nivo with it

**Comment by J4v4Scr1pt** [comment]
> Latest major version of Next.js (version 15) uses React 19 RC, so it's not going to be possible to use Nivo with it

Man.. this is no good..

**Comment by tarjep** [comment]
I was upgrading to Next 15 and by effect React 19, and ended up being blocked by this issue as well.

**Comment by jonasbuechel** [comment]
I'm about to use Nivo but also like to upgrade to Next 15 / React 19 the next few weeks. Is there a work-around for the moment?

**Comment by J4v4Scr1pt** [comment]
And React 19 is now stable...
https://react.dev/blog/2024/12/05/react-19

**Comment by furkanonb** [comment]
```
'ResponsiveBar' cannot be used as a JSX component.
  Its type '<RawDatum extends BarDatum>(props: ResponsiveBarSvgProps<RawDatum>) => Element' is not a valid JSX element type.
    Type '<RawDatum extends BarDatum>(props: ResponsiveBarSvgProps<RawDatum>) => Element' is not assignable to type '(props: any) => ReactNode'.
      Type 'Element' is not assignable to type 'ReactNode'.
        Property 'children' is missing in type 'ReactElement<any, any>' but required in type 'ReactPortal'.
```

I get a TypeScript error after upgrading React to 19. Is anyone else experiencing the same issue?

**Comment by furkanonb** [comment]
> ```
> 'ResponsiveBar' cannot be used as a JSX component.
>   Its type '<RawDatum extends BarDatum>(props: ResponsiveBarSvgProps<RawDatum>) => Element' is not a valid JSX element type.
>     Type '<RawDatum extends BarDatum>(props: ResponsiveBarSvgProps<RawDatum>) => Element' is not assignable to type '(props: any) => ReactNode'.
>       Type 'Element' is not assignable to type 'ReactNode'.
>         Property 'children' is missing in type 'ReactElement<any, any>' but required in type 'ReactPortal'.
> ```
> 
> I get a TypeScript error after upgrading React to 19. Is anyone else experiencing the same issue?

I solved my issue by adding it to the global.d.ts file.

```
import type { JSX as Jsx } from 'react/jsx-runtime';

declare global {
  namespace JSX {
    type ElementClass = Jsx.ElementClass;
    type Element = Jsx.Element;
    type IntrinsicElements = Jsx.IntrinsicElements;
  }
}
```

The reason is;
_A long-time request is to remove the global JSX namespace from our types in favor of React.JSX. This helps prevent pollution of global types which prevents conflicts between different UI libraries that leverage JSX._

[https://react.dev/blog/2024/04/25/react-19-upgrade-guide#the-jsx-namespace-in-typescript](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#the-jsx-namespace-in-typescript)

**Comment by J4v4Scr1pt** [comment]
I have updated my project to NextJS 15 and React 19. Using overrides and all seems to work just fine!
But I only have these Nivo dependencies:

```
	"@nivo/bar": "0.88.0",
	"@nivo/core": "0.88.0",
	"@nivo/pie": "0.88.0",
```

```
"pnpm": {
		"overrides": {
			"@types/react": "19.0.1",
			"@types/react-dom": "19.0.2"
		}
	}
```

**Comment by espetro** [comment]
Hey 👋 I'd like to support with enabling React 19 support in the project. How can we track the todos left to reach it?

**Comment by tohearen** [comment]
I'm also interested in using Nivo with Next 15+ / React 19+. Subscribed to this thread.

**Comment by tohearen** [comment]
I switched to @visx/visx@3.13.2-alpha.0 for now. It has proper support for React 19.

**Comment by plouc** [comment]
As someone reported it worked with overrides, I could increase the range of allowed version for now.

**Comment by J4v4Scr1pt** [comment]
> Are tooltips working for you folks? They do not seem to work for me (default or custom) migrating to react 19, nextjs 15 and nivo 0.88 (coming from 0.87) without throwing any error in Sunburst, Line and CirclePacking

Yes, it works for me, I use custom and regular tooltips. I also use custom layers on the ResponsivePie chart. You can see above the charts I'm using and versions.

**Comment by nass600** [comment]
Right, I was able to test it in isolated environments and the issue seems to be regarding nextjs 15. 

Can’t figure out exactly what is happening though.
Tried to disable SSR in several ways but it just doesn’t render the tooltips and leaves zero warnings or issues behind

**Comment by aclaghorn** [comment]
@nass600 I also have issues with tooltips not rendering. I do have the crosshairs render and follow the mouse cursor, but the tooltips are completely missing.

**Comment by aclaghorn** [comment]
@nass600 I was able to get tooltips working! I had to simply add the `@nivo/tooltip` package to my package.json:

```
"dependencies": {
    "@nivo/core": "^0.88.0",
    "@nivo/line": "^0.88.0",
    "@nivo/tooltip": "^0.88.0",
}
```

**Comment by musetraine** [comment]
maybe someone can share a plan for update it to react 19v, right now I have peer react@">= 16.14.0 < 19.0.0" from @nivo/core@0.88.0 and what to do next?

**Comment by JoseDRojasA** [comment]
@nivo/heatmap has also the same issue 🤔 
```
jose@MacBook project% npm i @nivo/heatmap
npm error code ERESOLVE
npm error ERESOLVE unable to resolve dependency tree
npm error
npm error While resolving: kosmo-gateway@0.0.64
npm error Found: react@19.0.0
npm error node_modules/react
npm error   react@"19.0.0" from the root project
npm error
npm error Could not resolve dependency:
npm error peer react@">= 16.14.0 < 19.0.0" from @nivo/heatmap@0.88.0
npm error node_modules/@nivo/heatmap
npm error   @nivo/heatmap@"*" from the root project
npm error
npm error Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
npm error
npm error
npm error For a full report see:
npm error /Users/jose/.npm/_logs/2025-01-21T02_38_33_653Z-eresolve-report.txt
npm error A complete log of this run can be found in: /Users/jose/.npm/_logs/2025-01-21T02_38_33_653Z-debug-0.log
```

**Comment by fstivicic** [comment]
We were very excited to use Nivo in our project. It looks great.
But we choose not to after discovering that it falls on back of one person. 
@plouc is there specific reason that this is the case? I see there is a lot of will for support for Nivo, like @espetro for this ticket.

**Comment by plouc** [comment]
@fstivicic, in the past, one person significantly contributed to the project but eventually had to step away due to other obligations and interests. Since then, finding help has been quite challenging.

I’d love to get more contributors, but over the past few years, I've been busy with work and personal commitments, making it difficult to properly maintain the library.

One of the biggest hurdles is timing—when people offer to help, they often need clear guidance on how to tackle specific features or projects, which is completely understandable. However, I rarely have the bandwidth to provide timely direction.

Ideally, I need help from someone already familiar with the codebase and infrastructure—someone who could assist with triaging issues, planning, and overall project management.

There are still open questions about supporting React 19. Can we maintain a single codebase across multiple React versions? Which tools (Storybook, testing frameworks, Gatsby, etc.) need upgrades? How should we test different versions effectively?

Lastly, the project once received small donations via Open Collective, which served as a bit of extra motivation for potential maintainers. However, contributions have nearly stopped, making it even harder to attract help.

**Comment by tjg** [comment]
Might organizing a bounty/funding be worthwhile, @markedwards? It could be much cheaper than migration for your company and others.

**Comment by shridey** [comment]
You guys are amazing yr .. I feel grateful to witness such a sweet community of developers discussing on a particular issue... I'm genuinely thankful for this... Coming to the issue, I've recently migrated our company's product from CRA to Vite, and while doing that, I upgraded all the dependencies to their latest versions, except for React because Nivo doesn't support its latest version... So this issue is relevant to me as well... Currently, we're using React v18.2.0... Hoping to have the support for v19... It would really be great..

**Comment by nohira** [comment]
Hi, I'm using nextjs 15 and react 19 and I'm overriding with `--legacy-peer-deps` and it seems to work pretty well. I'm using core, line, stream, radial-bar, radar and I had no issues so far

**Comment by KissBalazs** [comment]
Even though our application technically runs with the npm install --force workaround, we’re unfortunately experiencing test failures specifically related to components using nivo (e.g. "TypeError: Cannot read properties of undefined (reading 'ReactCurrentDispatcher')"). This has led us to temporarily revert back to React 18 to ensure test stability.

While we could mock the nivo components in our tests, they play a crucial role in our application, and testing their rendered output is important for us to ensure correctness. Given this and the current lack of momentum around updates for the library, we’re seriously considering migrating away from nivo.

We truly appreciate the work that has gone into this project, and we fully understand how challenging open-source maintenance can be with limited contributors. That said, being unable to move past React 18 is becoming increasingly limiting for our toolchain, especially as other parts of our stack move forward. Sadly, the GitHub discussions and Discord channels also seem to have quieted down recently, which makes it harder to navigate these issues.

I am sharing this feedback with all respect and in the hope that it helps highlight some of the pain points faced by projects relying on nivo.

**Comment by plouc** [comment]
I've extended the react version range in v0.89.0, but there's another [issue](https://github.com/pmndrs/react-spring/issues/2341) with `@react-spring` which has `"react": "^16.8.0 || ^17.0.0 || ^18.0.0"` in its `peerDependencies`.

Next, I'll to remove prop-types and default props from all packages.

**Comment by plouc** [comment]
> Even though our application technically runs with the npm install --force workaround, we’re unfortunately experiencing test failures specifically related to components using nivo (e.g. "TypeError: Cannot read properties of undefined (reading 'ReactCurrentDispatcher')"). This has led us to temporarily revert back to React 18 to ensure test stability.
> 
> While we could mock the nivo components in our tests, they play a crucial role in our application, and testing their rendered output is important for us to ensure correctness. Given this and the current lack of momentum around updates for the library, we’re seriously considering migrating away from nivo.
> 
> We truly appreciate the work that has gone into this project, and we fully understand how challenging open-source maintenance can be with limited contributors. That said, being unable to move past React 18 is becoming increasingly limiting for our toolchain, especially as other parts of our stack move forward. Sadly, the GitHub discussions and Discord channels also seem to have quieted down recently, which makes it harder to navigate these issues.
> 
> I am sharing this feedback with all respect and in the hope that it helps highlight some of the pain points faced by projects relying on nivo.

@KissBalazs, the error you mention seems to be due to different versions of react being used, so I'd recommend to use `overrides` for example. I have a nextjs project setup locally which runs fine, using client components.

**Comment by dagatsoin** [comment]
@plouc @florence-wolfe 
Hello, yes, the beta for the support for React 19 of react-spring is released v10.0.0-beta.0
No bug so far, but more we test, more we consolidate. Any feedbacks will be appreciated.

**Comment by joealden** [comment]
@plouc FYI https://github.com/pmndrs/react-spring/releases/tag/v10.0.0

**Comment by plouc** [comment]
@joealden, thx for the heads-up! And thank your hard work @dagatsoin!
I'm currently working on a pretty big PR to improve responsive charts, once I'm done with that I'll try to upgrade react-spring then.

**Comment by ImLunaHey** [comment]
all of the [recently added forwardRefs](https://github.com/plouc/nivo/pull/2757#issuecomment-2886202362) should also be removed for better react 19 support.

**Comment by plouc** [comment]
@ImLunaHey, I don't think so, `ref` support is really important, and I don't see why that's an issue for React 19, I still want to support older versions which don't support accessing `ref` directly as a prop.
They mention that they'll deprecate it at some point, not in React 19 as I understand it.

**Comment by plouc** [comment]
With the latest version of `react-spring`, there's no need for `overrides` anymore.

**Comment by herrherrmann** [comment]
@plouc Should the info from [your comment](https://github.com/plouc/nivo/issues/2618#issuecomment-2651277048) perhaps be included somewhere more prominently, e.g. in the readme file of the repo? It seems like a quite important information for people to keep in mind (the slow maintenance pace and potential stop of the project anytime soon).

Background: I’ve been doing some chart library research today, and Nivo once again landed on my list since I saw the new-ish release of version 0.99.0 (9 months ago). However, the situation described above does not inspire a lot of confidence for the future of the project (especially if React releases pick up pace again and might break backwards compatibility again).

What do you think – Would a small disclaimer make sense? It could also help attract new maintainers or even donations if people are more aware of the issue (since I’m sure there are many happy Nivo users today).
